File: Array&String/Sorting/CPP/maximize_sum_of_array_after_k_negations.cpp
#include <algorithm>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
public:
    int largestSumAfterKNegations(vector<int>& nums, int k) {
        vector<int> ans, pos;
        for (int x : nums) {
            if (x < 0) ans.push_back(x);
        }
        sort(ans.begin(), ans.end());
        for (int x : nums) {
            if (x >= 0) pos.push_back(x);
        }
        int n = ans.size();
        for (int i = 0; i < min(k, n); i++) {
            ans[i] = -ans[i];
        }
        k -= min(k, n);
        ans.reserve(ans.size() + pos.size());
        ans.insert(ans.end(), pos.begin(), pos.end());
        sort(ans.begin(), ans.end());
        if (k && k % 2)  ans[0] = -ans[0];
        return accumulate(ans.begin(), ans.end(), 0);
    }
};

File: Array&String/Sorting/maximize_sum_of_array_after_k_negations.py
class Solution:
    def largestSumAfterKNegations(self, nums: list[int], k: int) -> int:
        smalls = sorted([x for x in nums if x < 0])
        pos = [x for x in nums if x >= 0]
        n = len(smalls)
        for i in range(min(k, n)):
            smalls[i] = -smalls[i]
        k -= min(k, n)
        smalls = sorted(smalls + pos)
        if k and k % 2: smalls[0] = -smalls[0]
        return sum(smalls)


    # shorter
    def largestSumAfterKNegations(self, A: list[int], K: int) -> int:
        A.sort()
        i = 0
        while i < len(A) and i < K and A[i] < 0:
            A[i] = -A[i]
            i += 1
        return sum(A) - (K - i) % 2 * min(A) * 2
